File: bindings/java/examples/BasicMessaging.java
/** Basic messaging: create identity, create context, send and receive messages. */

package com.limn.scp.examples;

import com.limn.scp.Context;
import com.limn.scp.Identity;
import com.limn.scp.Types.Message;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Flow;

public class BasicMessaging {
    public static void main(String[] args) throws Exception {
        // Create two identities
        var alice = Identity.create("platform").join();
        var bob = Identity.create("platform").join();
        System.out.println("Alice DID: " + alice.did());
        System.out.println("Bob DID: " + bob.did());

        // Alice creates a context
        var ctxAlice = Context.create(alice, Map.of(
            "ceiling", List.of("msg:send", "msg:receive"),
            "ttl", 3600,
            "governance", "single_admin"
        )).join();
        System.out.println("Context ID: " + ctxAlice.contextId());

        // Bob joins the context
        var ctxBob = Context.join(bob, ctxAlice.contextId()).join();

        // Alice sends a message
        ctxAlice.send("Hello Bob, this is Alice".getBytes()).join();

        // Bob receives it
        ctxBob.receive().subscribe(new Flow.Subscriber<>() {
            private Flow.Subscription subscription;

            @Override
            public void onSubscribe(Flow.Subscription s) {
                this.subscription = s;
                s.request(1);
            }

            @Override
            public void onNext(Message msg) {
                System.out.println("Bob received from " + msg.senderDid() + ": " + new String(msg.content()));
                subscription.cancel();
            }

            @Override
            public void onError(Throwable t) { t.printStackTrace(); }

            @Override
            public void onComplete() {}
        });

        // Cleanup
        ctxBob.close();
        ctxAlice.close();
        alice.close();
        bob.close();
    }
}

File: bindings/go/examples/basic_messaging.go
// Basic messaging: create identity, create context, send and receive messages.
package main

import (
	"fmt"
	"log"

	scp "github.com/limn/scp-go"
)

func main() {
	if err := scp.Init(); err != nil {
		log.Fatal(err)
	}
	defer scp.Shutdown()

	// Create two identities
	alice, err := scp.NewIdentity("platform")
	if err != nil {
		log.Fatal(err)
	}
	bob, err := scp.NewIdentity("platform")
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println("Alice DID:", alice.DID())
	fmt.Println("Bob DID:", bob.DID())

	// Alice creates a context
	ctxAlice, err := scp.NewContext(alice, scp.ContextParams{
		Ceiling:    []string{"msg:send", "msg:receive"},
		TTL:        3600,
		Governance: "single_admin",
	})
	if err != nil {
		log.Fatal(err)
	}
	defer ctxAlice.Close()
	fmt.Println("Context ID:", ctxAlice.ContextID())

	// Bob joins the context
	ctxBob, err := scp.JoinContext(bob, ctxAlice.ContextID())
	if err != nil {
		log.Fatal(err)
	}

	// Alice sends a message
	if err := ctxAlice.Send([]byte("Hello Bob, this is Alice")); err != nil {
		log.Fatal(err)
	}

	// Bob receives it
	msg := <-ctxBob.Receive()
	fmt.Printf("Bob received from %s: %s\n", msg.SenderDID, msg.Content)

	// Cleanup
	ctxBob.Leave()
}
